### PR TITLE
[Persistence] Base Tx Indexer Implementation (Issue#151)

### LINKED ISSUE #151: [Persistence] Base transaction indexer
# Objective
- **CREATE** an `indexer` for arbitrary transaction bytes
- **USE** `protobuf3` (already used throughout the project) for encoding and a `filesystem db` for storage
- **INDEX** each transaction for `<sha3>hash`, `height`, `sender`, `recipient (if applicable)`
- **EXPOSE** an `interface` for write and read
- **TEST** create `unit`, `fuzz`, and `integration` tests

**Example interface**
 ```
type TxIndexer interface {  // TxIndexer interface defines methods to index and query transactions.

	// Index analyzes, indexes and stores a single transaction.
	Index(result *types.TxResult) error

	// GetByHash returns the transaction specified by hash or nil if the transaction is not indexed
	GetByHash(hash []byte) (*types.TxResult, error)

	// GetByHeight returns the transactions specified by height or nil if there are no transactions at that height
	GetByHeight(height int64) ([]*types.TxResult, error)

	// GetBySender returns the transactions signed by *sender* may be ordered descending/ascending
	GetBySender(sender address, descending bool) ([]*types.TxResult, error)

	// GetByRecipient returns the transactions *sent to address* may be ordered descending/ascending
	GetByRecipient(rec address, descending bool) ([]*types.TxResult, error)

	// DeleteFromHeight method added for pruning/rollback
	DeleteFromHeight(height int64) error
}

...
types/

type TxResult interface {
  Tx() Transaction        // the transaction object
  Height() int64          // height it was sent
  Index() int8            // which index it was within the block-transactions
  GetResultCode() int64   // 0 is no error, otherwise corresponds to error object code
  GetError() Error        // can be nil
  Signer() address        // get the address who signed
  Recipient() address     // can be nil
  MessageType() string    // corresponds to type of message (Ex. validator-stake, app-unjail, node-stake) etc.
}

type Transaction interface {
   // To be defined / for now it's just []byte
}

 ```

# Origin Document

Thus transaction storage (indexing) is a fundamental component of any blockchain system. For M1, this is *P0*

# Goals / Deliverables
- [ ] Code complete implementation 
- [ ] README and CHANGELOG for the indexer

## General issue checklist
- [ ] Update the appropriate CHANGELOG
- [ ] Update the README
- [ ] If applicable, update the source code tree explanation
- [ ] If applicable, add or update a state, sequence or flowchart diagram using [mermaid](https://mermaid-js.github.io/mermaid/)
- [ ] Update any relevant global documentation & references
- [ ] Document small issues / TODOs along the way

---

Creator: @andrewnguyen22

### PR DESCRIPTION
**Pull request by andrewnguyen22**
## Description
Transaction storage (indexing) is a fundamental component of any blockchain system. For M1, this is P0

closes #151

## Type of change
Please mark the options that are relevant.

- [x] New feature (non-breaking change which adds functionality)

## How Has This Been Tested?
- [x] Unit tests

___REPLACE_ME_: Describe the tests and that you ran to verify your changes. If applicable, provide steps to reproduce. Bonus points for images and videos or gifs._

- [x] `make test_all`
- [x] [LocalNet](https://github.com/pokt-network/pocket/blob/main/docs/development/README.md)

## Checklist
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have tested my changes using the available tooling

### CODE REVIEW

**Comment by andrewnguyen22** [comment]
Follow up created to integrate the indexer into the bus

File: shared/types/indexer.go
Comment (Olshansk) [inline shared/types/indexer.go:237]: For my personal learning and understanding, why are we using the `rune` type for the `prefix`?

File: shared/types/indexer_test.go
Comment (Olshansk) [inline shared/types/indexer_test.go:12]: I'm a fan of having complex unit tests testing a variety of different complex cases, but since this is a standalone package that we likely won't revisit in the future (unless there's a bug or new feature), can you please add behaviour specific unit tests

Here is a non-exhaustive list just as a starting point/reference: 
```go
func TestTransactionIndexerIndexByHeight(t * testing.T)
func TestTransactionIndexerIndexByHash(t * testing.T)
func TestTransactionIndexerIndexBySender(t * testing.T)
func TestTransactionIndexerIndexByRecipient(t * testing.T)
...

func TestTransactionIndexerIndexByHeightOnly(t * testing.T)
func TestTransactionIndexerIndexBySenderOnly(t * testing.T)
func TestTransactionIndexerIndexByRecipientOnly(t * testing.T)
...
```

File: shared/types/indexer.go
Comment (Olshansk) [inline shared/types/indexer.go:210]: Should this be an error? I don't think the recipient should ever be nill.

Add a unit test for this case too

File: shared/types/indexer_test.go
Comment (Olshansk) [inline shared/types/indexer_test.go:92]: Should we have an enum for this or a set of constants we reference?

File: shared/types/indexer_test.go
Comment (Olshansk) [inline shared/types/indexer_test.go:97]: Add a `require.NoError`

File: shared/types/proto/tx.proto
Comment (Olshansk) [inline shared/types/proto/tx.proto:11]: Should we have a TODO to have a type for errors or result_code?

File: shared/types/indexer_test.go
Comment (Olshansk) [inline shared/types/indexer_test.go:114]: for my understanding, why are we using `rune` instead of `byte` here?

File: shared/types/indexer_test.go
Comment (Olshansk) [inline shared/types/indexer_test.go:112]: Add a comment: E.g. `Generates a random alphanumeric sequence of exactly 50 characters`

File: shared/types/indexer_test.go
Comment (Olshansk) [inline shared/types/indexer_test.go:101]: Given some of the hardcoded values, we should add a comment: E.g. `Returns an error 25% of the time`

**Review by Olshansk — CHANGES_REQUESTED** [review]
@andrewnguyen22 Did a first round of review on this. Overall, all the business logic is there and it looks like it works, but I left a few comments that I think will make it more approachable to future maintainers and readers of the code (e.g. adding comments for implicit business logic in unit test helpers) as well as the codebase grows (e.g. putting the indexer in a separate package / subdirectory).

Some of the comments I left were very short and quick (and potentially repetitive) so feel free to ask for additional clarification if its needed.

File: persistence/kvstore/kvstore.go
Comment (andrewnguyen22) [reply persistence/kvstore/kvstore.go:9]: That's fine, there are multiple instances of this throughout the codebase where it's done like this. It'll probably need a cleanup issue

File: persistence/kvstore/kvstore.go
Comment (andrewnguyen22) [reply persistence/kvstore/kvstore.go:78]: Views is a badger specific thing. Will improve comment

File: shared/types/indexer.go
Comment (andrewnguyen22) [reply shared/types/indexer.go:6]: I'll run `gofmt` in the meantime

File: shared/types/indexer.go
Comment (andrewnguyen22) [reply shared/types/indexer.go:26]: Time put in the blockchain; `Block/Index`

File: shared/types/indexer.go
Comment (andrewnguyen22) [reply shared/types/indexer.go:39]: Yes

File: shared/types/indexer.go
Comment (andrewnguyen22) [reply shared/types/indexer.go:56]: I don't like a separate package for the indexer as it's simple and has a specific usecase 

When I move this to utility, I can put it in the Utility README. 

Does that work?

File: shared/types/proto/tx.proto
Comment (andrewnguyen22) [reply shared/types/proto/tx.proto:13]: It's just the address type for signer / recipient

File: shared/types/indexer.go
Comment (andrewnguyen22) [reply shared/types/indexer.go:81]: On goland it complains because all of the recievers don't have the same name. 

(x ...) comes from protobuf generator

File: shared/types/indexer.go
Comment (andrewnguyen22) [reply shared/types/indexer.go:85]: Yes, anytime we need to unmarshal `proto` bytes we can use this function

File: shared/types/indexer.go
Comment (andrewnguyen22) [reply shared/types/indexer.go:1]: See objection above

File: shared/types/indexer.go
Comment (andrewnguyen22) [reply shared/types/indexer.go:51]: Are you referring to our deployment? Just because we have it there doesn't mean it doesn't have a place in the codebase during development. These strict checks saved me several times in the past from breaking interfaces

File: shared/types/indexer.go
Comment (andrewnguyen22) [reply shared/types/indexer.go:36]: It's the result of playing a transaction agains the blockchain.

File: shared/types/indexer.go
Comment (andrewnguyen22) [reply shared/types/indexer.go:90]: TxResult is an interface which is a pointer by default.

A pointer to an interface ends up being a pointer to a pointer of a structure.

I'd rather not do that

File: shared/types/indexer.go
Comment (andrewnguyen22) [reply shared/types/indexer.go:176]: I can do that but I'd need to return the error object instead of assigning it to the already created one

File: shared/types/indexer.go
Comment (andrewnguyen22) [reply shared/types/indexer.go:237]: `rune` is a single unicode character which is like a `char` of the string

File: shared/types/indexer_test.go
Comment (andrewnguyen22) [reply shared/types/indexer_test.go:12]: I don't understand. I followed the same testing framework like the Consensus Module where I test public functions covering happy/bad path use cases.

I need more detail on this before I arbitrarily start increasing the number of tests when all of these are already covered

File: shared/types/indexer.go
Comment (andrewnguyen22) [reply shared/types/indexer.go:210]: There are many transactions that don't have a recipient. In fact, there is only one transaction type that has a recipient `send`

File: shared/types/indexer_test.go
Comment (andrewnguyen22) [reply shared/types/indexer_test.go:92]: As asked above, TODO is added

If you want another specifically for testing, I can also add that

File: shared/types/proto/tx.proto
Comment (andrewnguyen22) [reply shared/types/proto/tx.proto:11]: No, actually considering our block correctness design, both of these should be deprecated shortly. Will add TODO for that

Comment (andrewnguyen22) [reply shared/types/proto/tx.proto:11]: We have type for errors and code in the utility package. Will use those when moved

File: shared/types/indexer_test.go
Comment (andrewnguyen22) [reply shared/types/indexer_test.go:114]: Generally speaking, `rune` is regarded as the `char` primitive of a string. In this case, I think byte would also be acceptable since it's ascii chars

File: persistence/kvstore/kvstore.go
Comment (Olshansk) [reply persistence/kvstore/kvstore.go:9]: It was the only such addition in this PR, so just did it myself.

File: shared/types/indexer.go
Comment (Olshansk) [reply shared/types/indexer.go:11]: Changed it to `// TECHDEBT(andrew): Move this out of shared and alongside the mempool.`

File: shared/types/indexer.go
Comment (Olshansk) [reply shared/types/indexer.go:42]: Updated the comment to `// which index it was within the block-transactions; ordered by when the proposer received it in the mempool`

File: shared/types/proto/tx.proto
Comment (Olshansk) [inline shared/types/proto/tx.proto:1]: Thought on renaming this file to `transaction_indexer.proto` so it'll contain everything related to this business logic?

Simply `tx.proto` could be unclear to new readers of the coebase.

File: shared/types/proto/tx.proto
Comment (Olshansk) [inline shared/types/proto/tx.proto:6]: I add some additional comments below. PTAL

File: shared/types/proto/tx.proto
Comment (Olshansk) [reply shared/types/proto/tx.proto:13]: Can we rename the params to `signer_addr` and `recipient_addr`?

File: shared/modules/telemetry_module.go
Comment (Olshansk) [inline shared/modules/telemetry_module.go:1]: What were the changes in this file? Just formatting?

File: shared/types/indexer_test.go
Comment (Olshansk) [reply shared/types/indexer_test.go:12]: > I followed the same testing framework like the Consensus Module where I test public functions covering happy/bad path use cases

I don't think I did a good job on the consensus module level tests when we first implemented it, and definitely think we should add finer grained ones as well, ina addition to the module-level ones.

> I need more detail on this before I arbitrarily start increasing the number of tests when all of these are already covered

I was thinking something similar to what we have for the actor level tests in persistence and utility. Example:

We have unit tests like so:

```
TestInsertFishermanAndExists
TestUpdateFisherman
TestGetFishermenReadyToUnstake
TestGetFishermanStatus
TestGetFishermanPauseHeightIfExists
TestSetFishermanPauseHeightAndUnstakeLater
```

In addition to end-to-end tests like so:

```
FuzzFisherman
```

I think it's important that we do this now and have them in place before merging this in.

File: shared/types/indexer.go
Comment (Olshansk) [reply shared/types/indexer.go]: I believe the cost of having it be a separate package right now is low and will improve the health of the codebase. Added a `DISCUSS_IN_THIS_COMMIT` but feel free to change to `TODO_IN_THIS_COMMIT(olshansky)` if you'd like for me to take it on.

File: shared/types/indexer.go
Comment (Olshansk) [reply shared/types/indexer.go:1]: Adding a `DISCUSS_IN_THIS_COMMIT`

File: shared/types/indexer.go
Comment (Olshansk) [inline shared/types/indexer.go:81]: Please add a reference link

File: shared/types/indexer.go
Comment (Olshansk) [inline shared/types/indexer.go:82]: Add a `INVESTIGATE` prefix to the comment

File: shared/types/indexer.go
Comment (Olshansk) [reply shared/types/indexer.go:94]: Makes sense.

**Review by Olshansk — CHANGES_REQUESTED** [review]
Left another round of PR comments and pushed some minor changes myself (PTAL a those)

File: shared/types/indexer.go
Comment (andrewnguyen22) [reply shared/types/indexer.go:84]: Research topic in identified in #178

File: shared/types/indexer.go
Comment (andrewnguyen22) [reply shared/types/indexer.go:80]: I don't think this is needed because it's included in the source code dependency README.md

File: shared/indexer/indexer_test.go
Comment (Olshansk) [inline shared/indexer/indexer_test.go:45]: Why is the index the number of results? Ditto bellow.

Is it meant to represent its position?

I saw the switch statement below but still find it hard to follow.

File: shared/indexer/indexer_test.go
Comment (Olshansk) [inline shared/indexer/indexer_test.go:190]: You can write `require.Equal(bz, bz2)`. If it's not equal, it'll output the diff too.

Will help remove `require.True` elsewhere as well.

File: shared/indexer/indexer_test.go
Comment (Olshansk) [inline shared/indexer/indexer_test.go:137]: What do the numbers 2 and 1 mean here?

**Review by Olshansk — CHANGES_REQUESTED** [review]

File: shared/indexer/indexer_test.go
Comment (andrewnguyen22) [reply shared/indexer/indexer_test.go:210]: I don't agree - this is testing code specific for the indexer. Also, it's utility specific. So it should be in the Utility mod if anything (where the indexer should live)

File: shared/indexer/proto/transaction_indexer.proto
Comment (andrewnguyen22) [reply shared/indexer/proto/transaction_indexer.proto:9]: I like index better IMO.

I've not seen a blockchain use txPosition

File: shared/types/indexer.go
Comment (Olshansk) [reply shared/types/indexer.go]: `These strict checks saved me several times in the past from breaking interfaces`

+1 agreed about having them here (and elsewhere). 

I do not remember the context of my original comment so resolving.

File: shared/types/indexer.go
Comment (Olshansk) [inline shared/types/indexer.go]:
```suggestion
        // Most transactions do not have a recipient, so this is a common and expected happy path.
	if recipient == "" {
```

File: shared/types/indexer.go
Comment (Olshansk) [reply shared/types/indexer.go]: Noted. Adding a comment suggestion

File: shared/types/indexer_test.go
Comment (Olshansk) [reply shared/types/indexer_test.go]: Noted. Found this online: https://www.callicoder.com/golang-basic-types-operators-type-conversion

**tl;dr `rune` is UTF8 and `char` is ASCII only.**

Feel free to resolve.

File: shared/indexer/proto/transaction_indexer.proto
Comment (Olshansk) [reply shared/indexer/proto/transaction_indexer.proto:9]: Sgtm, trust you on this.

Comment (Olshansk) [inline shared/indexer/proto/transaction_indexer.proto:9]:
```suggestion
  int32 index = 3;   // The order (i.e. position within a block) where the proposer included the transaction
```

File: shared/indexer/indexer_test.go
Comment (Olshansk) [reply shared/indexer/indexer_test.go:210]: Got it. 

Let's leave as is, and we can always revisit in the future if/when it comes up.

**Review by Olshansk — APPROVED** [review]
Thanks for all the changes @andrewnguyen22, in particular with the comments, tests and clarifications!

I left a couple of comment suggestions that you can just "accept", but otherwise LGTM

**Pull request merged**
